// gitbook-plugin-page-copyright/index.js
Date.prototype.format = function(format) {
	var date = {
		"M+": this.getMonth() + 1,
		"d+": this.getDate(),
		"h+": this.getHours(),
		"m+": this.getMinutes(),
		"s+": this.getSeconds(),
		"q+": Math.floor((this.getMonth() + 3) / 3),
		"S+": this.getMilliseconds()
	};
	if (/(y+)/i.test(format)) {
		format = format.replace(RegExp.$1, (this.getFullYear() + '').substr(4 - RegExp.$1.length));
	}
	for (var k in date) {
		if (new RegExp("(" + k + ")").test(format)) {
			format = format.replace(RegExp.$1, RegExp.$1.length == 1 ? date[k] : ("00" + date[k]).substr(("" + date[k]).length));
		}
	}
	return format;
};

/**
 * [main module]
 * @type {Object}
 */
const pageFooter = module.exports = {
	/** Map of new style */
	book: {
		assets: './assets',
		css: [
			'footer.css'
		]
	},

	/** Map of hooks */
	hooks: {
		'page:before': function (page) {
			if (this.output.name != 'website') {
				return page;
			}

			/**
			 * [defaultOption: default option]
			 * @type {Object}
			 */
			const defaultOption = {
				'description': 'modified at',
				'signature': 'Skylor.min',
				'wisdom': 'Designer, Frontend Developer & overall web enthusiast',
				'format': 'YYYY-MM-dd hh:mm:ss',
				'copyright': 'Copyright &#169; skylor',
				'timeColor': '#666',
				'copyrightColor': '#666',
				'utcOffset': '8',
				'isShowQRCode': true,
				'baseUri': 'https://github.com/skyFi',
				'isShowIssues': true,
				'repo': 'skyFi/gitbook-plugin-page-footer',
				'issueNum': '8',
				'token': '',
				'style': 'normal'
			};
			/**
			 * [configOption: config option]
			 * @type {Object}
			 */
			const configOption = this.config.get('pluginsConfig')['page-copyright'];

			/** if users have its option, and then combine it with default options */
			if (configOption) {
			// @deprecated
			// if (this.options.pluginsConfig['page-copyright']) {
				for (var item in defaultOption) {
					/** special for copyright */
					// @deprecated
					// defaultOption[item] = this.options.pluginsConfig['page-copyright'][item] || defaultOption[item];
					if (item in configOption) {
						defaultOption[item] = configOption[item];
					}

					if (item === 'copyright') {
						defaultOption[item] += ' all right reserved';

						if (!configOption.noPowered) {
							defaultOption[item] += ', powered by <a href="https://github.com/skyFi" target="_blank">skyFi</a>';
						}
					}
				}
			}

			defaultOption.style = (defaultOption.style == 'normal' || defaultOption.style == 'symmetrical') ? defaultOption.style : 'normal';

			const htmlContents = ' \n\n<footer class="footer">' +
				'<div class="footer__container--' + defaultOption.style + '">' +
					'<div class="footer__description--' + defaultOption.style + '">' +
						'<p class="paragraph footer__author--' + defaultOption.style + '" style="color: #000 !important;">' + defaultOption.signature + '<sup class="super">&#174;</sup></p>' +
						'<p class="paragraph footer__quote--' + defaultOption.style + '" style="color: #000 !important;">' + defaultOption.wisdom + '</p>' +
						'<div class="footer__main--' + defaultOption.style + '">' +
							'<p class="paragraph footer__main__paragraph--' + defaultOption.style + ' copyright" style="color: ' + defaultOption.copyrightColor + ' !important;">' + defaultOption.copyright +  '</span>' +
							'<p class="paragraph footer__main__paragraph--' + defaultOption.style + ' footer__modifyTime--' + defaultOption.style + '" style="color: ' + defaultOption.timeColor + ' !important;">' +
								'<span style="color: #666 !important;">' + defaultOption.description + '</span>' +
								'\n{{ file.mtime | dateFormat("' + defaultOption.format + '", ' + defaultOption.utcOffset + ') }}\n' +
							'</p>' +
						'</div>' +
					'</div>' +
				'</div>' +
				'<div id="gitalk-container" class="box__issues"></div>' +
				'<link rel="stylesheet" href="../assets/gitalk.min.css"><script src="../assets/gitalk.min.js"></script>' +
				'<script>var gitalk = new Gitalk({clientID: "1dd340a7923ebd556673",clientSecret: "4981d6b240efb08d88e19b395d3b7ea12300e0ec",repo: "blog.fenotes.com",owner: "uncle-marion",admin: ["uncle-marion"],id: location.pathname});gitalk.render("gitalk-container");</script>' +
			'</footer>';

			/** add contents to the original content */
			page.content = page.content + htmlContents;

			return page;
		}
	},

	/** Map of new blocks */
	blocks: {},

	/** Map of new filters */
	filters: {
		dateFormat: function(d, format, utc) {
			var reservedDate = new Date(d);
			/** convert to UTC firstly */
			reservedDate = new Date(
				reservedDate.getUTCFullYear(),
				reservedDate.getUTCMonth(),
				reservedDate.getUTCDate(),
				reservedDate.getUTCHours(),
				reservedDate.getUTCMinutes(),
				reservedDate.getUTCSeconds()
			);
			reservedDate.setTime(reservedDate.getTime() + (!utc ? 8 : parseInt(utc)) * 60 * 60 * 1000);
			return reservedDate.format(format);
		},

		convertUri: function (d, baseUri) {
			return baseUri + this.output.toURL(d);
		},

		currentUri: function (d, baseUri) {
			if (this.output.name == 'website') {
				return pageFooter.createQRcode(baseUri + this.output.toURL(d), 15, 'Q');
			} else {
				return '';
			}
		},

		listRepo: function (d, token, format, utc, issueNum) {
			var content = '';
			return content;
		}
	},

	/**
	 * [test: tests function]
	 * @param  {[type]} configs [simulated configs]
	 * @return {[type]}        [description]
	 */
	test: function (configs) {
		/**
		 * [option: default option]
		 * @type {Object}
		 */
		const defaultOption = {
			'description': 'modified at',
			'signature': 'Aleen',
			'wisdom': 'More than a coder, more than a designer',
			'format': 'yyyy-MM-dd hh:mm:ss',
			'copyright': 'Copyright &#169; aleen42',
			'timeColor': '#666',
			'copyrightColor': '#666',
			'utcOffset': '8',
			'isShowQRCode': true,
			'baseUri': 'https://aleen42.gitbooks.io/personalwiki/content/',
			'isShowIssues': true,
			'repo': 'aleen42/PersonalWiki',
			'token': '',
			'style': 'normal',
			'noPowered': false,
		};

		/** if users have its option, and then combine it with default options */
		if (configs['page-copyright']) {
			// @deprecated
			// if (this.options.pluginsConfig['page-copyright']) {
			for (var item in defaultOption) {
				/** special for copyright */
				// @deprecated
				// defaultOption[item] = this.options.pluginsConfig['page-copyright'][item] || defaultOption[item];
				if (item in configs) {
					defaultOption[item] = configs[item];
				}

				if (item === 'copyright') {
						defaultOption[item] += ' all right reserved';

						if (!configOption.noPowered) {
							defaultOption[item] += ', powered by <a href="https://github.com/skyFi" target="_blank">skyFi</a>';
						}
					}
			}
		} else {
			defaultOption.copyright += ' all right reserved';
		}

		/**
		 * [htmlContents: to store html tags]
		 * @type {String}
		 */
		const issues = defaultOption.isShowIssues === true ? '\n{{ "' + defaultOption.repo + '" | listRepo("' + defaultOption.token + '") }}\n' : '';

		defaultOption.style = (defaultOption.style == 'normal' || defaultOption.style == 'symmetrical') ? defaultOption.style : 'normal';

		const htmlContents = ' \n\n<footer class="footer">' +
			'<div class="footer__container--' + defaultOption.style + '">' +
				'<div class="footer__description--' + defaultOption.style + '">' +
					'<p class="paragraph footer__author--' + defaultOption.style + '" style="color: #000 !important;">' + defaultOption.signature + '<sup class="super">&#174;</sup></p>' +
					'<p class="paragraph footer__quote--' + defaultOption.style + '" style="color: #000 !important;">' + defaultOption.wisdom + '</p>' +
					'<div class="footer__main--' + defaultOption.style + '">' +
						'<p class="paragraph footer__main__paragraph--' + defaultOption.style + ' copyright" style="color: ' + defaultOption.copyrightColor + ' !important;">' + defaultOption.copyright +  '</span>' +
						'<p class="paragraph footer__main__paragraph--' + defaultOption.style + ' footer__modifyTime--' + defaultOption.style + '" style="color: ' + defaultOption.timeColor + ' !important;">' +
							'<span style="color: #666 !important;">' + defaultOption.description + '</span>' +
							'\n{{ file.mtime | dateFormat("' + defaultOption.format + '", ' + defaultOption.utcOffset + ') }}\n' +
						'</p>' +
					'</div>' +
				'</div>' +
			'</div>' +
			'<div>' +
				issues +
			'</div>' +
		'</footer>';

		return htmlContents;
	},
};
